### PR TITLE
Add workaround to GPIO MCUX driver for non contiguous pins

### LINKED ISSUE #44391: tests-ci : peripheral: gpio: 1pin test Timeout
**Describe the bug**
1pin test is Timeout on zephyr-v3.0.0-1964-g199048b2ed45 on mimxrt1015_evk

see logs for details

**To Reproduce**
1. 
```
scripts/twister --device-testing --device-serial /dev/ttyACM0 -p mimxrt1015_evk  --sub-test peripheral.gpio
```
2. See error

**Expected behavior**
test pass

**Impact**

**Logs and console output**
```
*** Booting Zephyr OS build zephyr-v3.0.0-1964-g199048b2ed45  ***
Running test suite gpio_api_1pin_test
===================================================================
START - test_gpio_pin_configure_push_pull
Running test on port=GPIO_3, pin=21
ASSERTION FAIL [pin < config->mux_count] @ WEST_TOPDIR/zephyr/drivers/gpio/gpio_mcux_igpio.c:142
E: r0/a1:  0x00000004  r1/a2:  0x0000008e  r2/a3:  0x40184000
E: r3/a4:  0x60006131 r12/ip:  0x00000000 r14/lr:  0x60006269
E:  xpsr:  0x41000000
E: Faulting instruction address (r15/pc): 0x60008c6c
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x20000248 (test_gpio_pin_configure_push_pu)
E: Halting system

    
```

**Environment (please complete the following information):**
 - OS: (e.g. Linux )
 - Toolchain (e.g Zephyr SDK)
 - Commit SHA or Version used: zephyr-v3.0.0-1964-g199048b2ed45

### PR DESCRIPTION
**Pull request by danieldegrasse**
Some iMX RT parts have a non contiguous set of pins available on certain GPIO ports (for example, GPIO0_0-GPIO0_10 are defined, as well as GPIO0_20-GPIO0_31, but GPIO0_11-GPIO0_19 are absent). Add a workaround to the iMX RT gpio driver to correctly select the GPIO pinmux register in these cases.

This PR also fixes the GPIO active flags on the RT1024 and RT1170 EVK LEDs, as these were discovered to be incorrect in testing.

Fixes #44391

### CODE REVIEW

File: drivers/gpio/gpio_mcux_igpio.c
Comment (mmahadevan108) [inline drivers/gpio/gpio_mcux_igpio.c:162]: Should we retain the comment and the `assert `check?

File: drivers/gpio/gpio_mcux_igpio.c
Comment (hakehuang) [inline drivers/gpio/gpio_mcux_igpio.c:56]: how about rt1010?

File: drivers/gpio/gpio_mcux_igpio.c
Comment (danieldegrasse) [reply drivers/gpio/gpio_mcux_igpio.c:162]: The assert was moved earlier in the function, because we were using `cfg_idx` to access the pin mux array before this assert was actually done. Now, the assertion will be checked before we use `cfg_idx`

**Review by mmahadevan108 — DISMISSED** [review]

File: drivers/gpio/gpio_mcux_igpio.c
Comment (danieldegrasse) [reply drivers/gpio/gpio_mcux_igpio.c:56]: For GPIO5? Those pins seem to be contiguous: 
image

If you mean the gap in GPIO3 pins, that is handled in the next `else if` case

Comment (hakehuang) [reply drivers/gpio/gpio_mcux_igpio.c:56]: I did not see the RT1010 in hte else if only 1024 and 1021

**Review by dleach02 — DISMISSED** [review]

**Comment by dleach02** [comment]
@hakehuang I'll wait on your approval before merging

**Comment by mbolivar-nordic** [comment]
DNM both while we're waiting on @hakehuang but also because this rings a bell for me and I want to double check if there is a standard way to express this in devicetree.

**Comment by mbolivar-nordic** [comment]
> I want to double check if there is a standard way to express this in devicetree.

There is; `gpio-reserved-ranges`: https://github.com/torvalds/linux/blob/59250f8a7f3a60a2661b84cbafc1e0eb5d05ec9b/Documentation/devicetree/bindings/gpio/gpio.txt#L150

It seems like it would be a bit cleaner to express this in the .dtsi files for each target that has this restriction. You could save the entire value of the array property in ROM and then quickly check if a given GPIO is valid at runtime or something like that.

File: drivers/gpio/gpio_mcux_igpio.c
Comment (hakehuang) [inline drivers/gpio/gpio_mcux_igpio.c:60]: @danieldegrasse , I did not see SOC_MIMXRT1011 is checking here

**Review by hakehuang — COMMENTED** [review]
TEst PASS, but SOC_MIMXRT1015 need add the gpio3 checking. this case does not cover this port though.

**Comment by danieldegrasse** [comment]
@mbolivar-nordic I agree that using `gpio-reserved-ranges` would be a cleaner approach to use here, but is there a way to access the index of the reserved GPIO and the number of reserved GPIOs induvidually? Something similar to the way that an array of register properties will have both a size and reg property available in the devicetree. The other option is of course just to create an array out of the `gpio-reserved-ranges` property and index into it properly to obtain each pair of (`gpio-idx, num-rsvd`) values, but I wanted to check if there was a better method to use.

**Review by hakehuang — DISMISSED** [review]
1015 gpio3 need enhance, others looks good to me

File: drivers/gpio/gpio_mcux_igpio.c
Comment (danieldegrasse) [inline drivers/gpio/gpio_mcux_igpio.c:58]: @hakehuang This check should cover the RT1015 GPIO5

**Comment by dleach02** [comment]
@mbolivar-nordic, can you consider @danieldegrasse response to your point and if okay remove the DNM label you added?

**Comment by mbolivar-nordic** [comment]
I didn't have time to get to this by code freeze, so I'm removing the DNM

**Comment by mbolivar-nordic** [comment]
@danieldegrasse sorry for the slow response; I'm fine with this, but I can't merge as there seem to be conflicts. Could you please rebase?

**Comment by danieldegrasse** [comment]
@mbolivar-nordic Rebased to fix conflicts

**Review by mbolivar-nordic — DISMISSED** [review]
@dleach02 @mmahadevan108 could you ACK the rebase and merge?

**Review by dleach02 — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by mmahadevan108 — DISMISSED** [review]

**Comment by dleach02** [comment]
All I did was a rebase to try to get rid of the errors... seems to have worked

**Review by dleach02 — APPROVED** [review]

**Review by hakehuang — APPROVED** [review]

**Pull request merged**
